### DISCUSSION TITLE
How to get a freshly created edge (for e.g. storing in a database) ?

### DISCUSSION THREAD

**Discussion by wowtah** [xyflow/xyflow#4242] [Q&A]
When I create a new Edge in the UI, I want get the freshly created edge in code, so I can store it (for example). But I think I haven't found the correct way of doing this. 

My idea was to subscribe to the The onEdgesChange callback, which (according to https://reactflow.dev/api-reference/types/edge-change) has a EdgeAddChange<T> variant, which should contain the Edge item. Through the callback I get all kinds of changes to the edges, but never the Add-variant. (see my code below). 

Then I looked at the 'onConnect' callback, but this gives back a 'Connection', which doesn't contain the new Edge (https://reactflow.dev/api-reference/types/connection). 

So, my question is: **how can I get a freshly created Edge object?** 

Of course I don't want to do things like a before- and after comparison of the edges collection :)  

```
"use client";
import React, { useCallback } from "react";
import ReactFlow, { useNodesState, useEdgesState, addEdge, NodeChange, EdgeChange, Panel } from "reactflow";

import "reactflow/dist/style.css";

const initialNodes = [
  { id: "1", position: { x: 0, y: 0 }, data: { label: "1" } },
  { id: "2", position: { x: 0, y: 100 }, data: { label: "2" } },
];
const initialEdges = [{ id: "e1-2", source: "1", target: "2" }];

export default function App() {
  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);

  const onConnect = useCallback((params: any) => setEdges((eds) => addEdge(params, eds)), [setEdges]);

  // Function to log changes
  const onNodeChangeLogger = (changes: NodeChange[]) => {
    console.log("Node changes:", changes);
    onNodesChange(changes);
  };

  const onEdgeChangeLogger = (changes: EdgeChange[]) => {
    console.log("Edge changes:", changes);
    onEdgesChange(changes);
  };

  // Click handler to add a new node
  const addNode = () => {
    const newNode = {
      id: (Math.random() * 1000).toString(),
      position: {
        x: Math.random() * window.innerWidth,
        y: Math.random() * window.innerHeight,
      },
      data: { label: "New Node" },
    };
    setNodes((prevNodes) => prevNodes.concat(newNode));
  };

  return (
    <div style={{ width: "100vw", height: "100vh" }}>
      <ReactFlow nodes={nodes} edges={edges} onNodesChange={onNodeChangeLogger} onEdgesChange={onEdgeChangeLogger} onConnect={onConnect}>
        <Panel position="top-center">
          <button onClick={addNode}>Add Node</button>
        </Panel>
      </ReactFlow>
    </div>
  );
}


```

As a side note; I see the same behaviour when I add a new node. I see all kinds of changes to the nodes (position, selected, deleted) but never the NodeAddChange<T = any>, which should exist according to https://reactflow.dev/api-reference/types/node-change

**Comment by wmurphyrd** [comment]
Same issue here - never able to observe NodeAddChange. [Here's a sandbox](https://codesandbox.io/p/sandbox/pedantic-smoke-mr886s?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clzaf7osk00063b6rnfxlh58y%2522%252C%2522sizes%2522%253A%255B100%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clzaf7osk00023b6rbixz5pwd%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clzaf7osk00033b6r77ylu9i0%2522%257D%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clzaf7osk00053b6r1b0ieqy9%2522%257D%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clzaf7osk00023b6rbixz5pwd%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzaf7osj00013b6ru27e3ncy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clzafaus200173b6rc6h69jgq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A19%252C%2522startColumn%2522%253A17%252C%2522endLineNumber%2522%253A19%252C%2522endColumn%2522%253A17%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clzaf7osk00023b6rbixz5pwd%2522%252C%2522activeTabId%2522%253A%2522clzafaus200173b6rc6h69jgq%2522%257D%252C%2522clzaf7osk00053b6r1b0ieqy9%2522%253A%257B%2522id%2522%253A%2522clzaf7osk00053b6r1b0ieqy9%2522%252C%2522activeTabId%2522%253A%2522clzafg40x003o3b6qis0fhvjw%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clzaf7osk00043b6rrbkx129o%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522id%2522%253A%2522clzafg40x003o3b6qis0fhvjw%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clzaf7osk00033b6r77ylu9i0%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clzaf7osk00033b6r77ylu9i0%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) that shows the changes right after an addition contain only the "dimensions" change type (check the console after load)

**Reply by CSworenI** [reply]
The simple answer is `useState`:
```
		onConnect = params => {
			const source = nodes.find(n => n.id === params.source);
			params.type = source.type === 'decision' ? 'decision' : 'smoothstep';
			params.id = `e${params.source}-${params.target}`; // crypto.randomUUID();
			setEdges((eds) => addEdge(params, eds));
			setCurrentEdge(params);
		},
```
If you log `edges` from `useEdgesState` on the next addition, it will appear, but not in the current one.

I'm having the issue of nodes not being added to the `nodes` state of `useNodesState` and being unable to connect new nodes as a result. Have you run into that?

**Comment by Santoshsohani** [comment]
https://github.com/user-attachments/assets/f68037f1-e9dd-4255-b44f-666d41e63e6e

```
import React, { useEffect, useState } from 'react';
import {
  ReactFlow,
  useNodesState,
  useEdgesState,
  MiniMap,
  Controls,
  Background,
  Panel,
  addEdge
} from '@xyflow/react';
import '@xyflow/react/dist/style.css';

const initialNodes = [
  { id: '1', data: { label: 'Node 1' }, position: { x: 200, y: 0 } },
];

const initialEdges = [];

const ChangeToDisplay = () => {
  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);
  const [currentEdge, setCurrentEdge] = useState(null);
  const [count,setCount] = useState(2);

  const handleAddNode = () => {
    setNodes((nodes) => [
      ...nodes,
      { id: `${count}`, data: { label: `Node ${count}` }, position: { x: 0, y: 0 } },
    ]);
    setCount(count+1)
  }

  const onConnect = (params) => {
    setEdges((eds) => addEdge({ ...params, type: 'smoothstep' }, eds));
  };

  useEffect(() => {
    console.log("Nodes : ",nodes);
    console.log("Edges : ", edges);
  },[nodes,edges])

  return (
    <div className="w-[100vw] h-[100vh]">
      <ReactFlow
        nodes={nodes}
        edges={edges}
        onNodesChange={onNodesChange}
        onEdgesChange={onEdgesChange}
        onConnect={onConnect}
        fitView
      >
        <MiniMap />
        <Controls />
        <Background variant="dots" />
        <Panel position="top-right">
          <button onClick={handleAddNode} className='bg-purple-800 text-white font-bold py-2 px-4 rounded'>Add Node</button>
        </Panel>
      </ReactFlow>
    </div>
  );
};

export default ChangeToDisplay;
```

I hope this helps :)

**Reply by CSworenI** [reply]
@wowtah this should be the accepted answer. Working well for me.

**Reply by Santoshsohani** [reply]
> @wowtah this should be the accepted answer. Working well for me.

Glad it helped 😄
